### ISSUE TITLE
`Module::set_training` takes hard-to-use `&mut self`

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/candle#783]
Hello everybody,

I noticed that `Module::set_training` takes an `&mut self`. This is hard to use for a non `Clone` and `Copy` type such as `Linear`. Taking an `&mut Linear` may seem viable, but my application requires an `Rc<Linear>`, so there is no `&mut` when there are more than 1 strong refs.

While looking through the code, it looks like there are only 3 `impl Module` in `candle` itself right now, so any changes would not break much code yet. I propose changing the method from taking an `&mut self` to `&self`. The trainability state would be in a `Cell`, and could be updated with `Cell::set` through an `&self`. To avoid sentinel values and stay idiomatic, instead of using a `bool`, an `enum` should be used with a `Default` impl of `Trainable`.

In addition, would implementing `Clone` on `Linear` and `Tensor` be a viable alternative? It looks like it can because `Tensor` is an `Arc<Tensor_>`.

What are your thoughts on this change?

Thanks,
Eric

**Comment by LaurentMazare** [comment]
Ah sorry, this `set_training` was actually a mistake as @Narsil already pointed out offline and should have been removed earlier, I've now removed it so that people don't start relying on it.
To me the main downsides were:
- It's stateful, so requires this `mut` bit but also makes it easy to forget to set it and use some dropout at inference time for example.
- If we provide a default implementation, it's easy to forget to pass it to submodules, if we don't it's annoying to have to set it even when writing code that will only be used for inference - or some training code that doesn't use a different behavior (so pretty much any architecture that is trained without dropout).
Instead I would want to favor a more stateless approach, i.e. if you want to train a model with a different behavior compared to inference, you could use a trait like:
```rust
trait ModuleT {
    fn forward_t(&self, xs: &Tensor, is_training: bool) -> Result<Tensor>;
}
```
We don't provide it at the moment as it's a bit unclear whether having a `is_training: bool` is sufficient or whether we want to have some more generic parameter, we would have to think more about it.

The other way is obviously to not use any trait and rather have the `forward` method that use the arguments that you care about, this is what is done in all the examples where the `forward` method needs multiple tensors or additional arguments etc.

What was your use case for using the `Module` trait, is it something that you can do without?

**Comment by EricLBuehler** [comment]
Ok, that make sense. I am working on implementing LoRA for Candle: https://github.com/EricLBuehler/candle-lora, which means that the `Linear` needs to have a trainable state (so that the gradients are not calculated, and therefore there are fewer trainable params for the old, wrapped layer).

If `Linear` has no state,, then it seems like `LoRA` cannot be implemented. When someone calls `Optimizer::backward_step`, it must be ensured that this tensor's weights and biases are not updated. So, is there a way to do this without `Tensor` storing state?

Perhaps the `set_training` method should be added again, with `Tensor` storing state so that it causes a short-circuit when `Tensor::backward` is called? As described in my original comment, `set_training` should take an `&self`, and `Linear` should also implement it for convenience.

**Comment by LaurentMazare** [comment]
Backprop only happens for tensors that are marked as "variables" (so created via `candle::Var`), regular tensors that you would load from say a safetensor file would not be tracked for backprop unless you explicitely make a var out of them. Wouldn't that make it possible to achieve what you describe?
Also note that the `Linear` layer is a [very small amount of code](https://github.com/huggingface/candle/blob/3cd7e7b51dc7bf49215b136702f5bc3cd4642144/candle-nn/src/linear.rs#L42-L55), so if you want some specialized version of it you can definitely make one inside your model (and implement the traits that you want on it).

**Comment by EricLBuehler** [comment]
That make sense, but do all trainable parameters need to be wrapped in `candle::Var`?

It looks like `Tensor` implements a `backward` method, and that `Optimizer::backward_step` takes a `Tensor`.  I also see this:
```rust
if node.is_variable() {
    continue;
}
```
in `Tensor::backward`, so is it skipping any computations with a `Var`?

**Comment by LaurentMazare** [comment]
Yes all trainable parameters should be in a var, we backprop through tensors (hence the `Tensor::backward`), you can get the grad of a tensor as you noticed but this will only get populated if this tensor has a dependency on a var. The `continue` in the code in the var case is a technical detail that you shouldn't worry about.

**Comment by EricLBuehler** [comment]
Ok, thank you! Should I use `VarBuilder` to store my tensors in a `Var`? Or do I need to manually wrap with `Var`?

**Comment by LaurentMazare** [comment]
A `VarBuilder` based on a `VarMap` would be the recommended way to go, but you can certainly create your `Var` structs manually if you prefer, you can see how `VarMap` does it [here](https://github.com/huggingface/candle/blob/3cd7e7b51dc7bf49215b136702f5bc3cd4642144/candle-nn/src/var_map.rs#L61).

If your linear layer is using weights that are not `Var` they won't be trainable (and gradients won't be tracked), up to you to decide how you create these weights, they just have to be tensors. Among the instances of `VarBuilder`, only the `VarMap` version creates `Var`, all the others create non-trainable `Tensor` (e.g. when loading from safetensor/npz/ggml/...).

**Comment by EricLBuehler** [comment]
Ok, that makes sense. I am writing a library, so I do not control what is passed. How do I ensure the weights and biases of the `Linea`r layer are not variables? Can I get their "name" and then try to remove them from the `VarBuilder`?

**Comment by EricLBuehler** [comment]
Is there a way to make sure that the old `Linear` layer's weights and biases are not variables? Or is this something that can not be done currently?

**Comment by LaurentMazare** [comment]
I don't think so, you may want to create your own `NonTrainableLinear` or equivalent in your crate if it's important to your use case.

**Comment by EricLBuehler** [comment]
Would that be effective? I take the old linear as `old: Arc<Linear>`. In my `forward` method, I call `.forward` on the `old` linear, and do the required `LoRA` computations. The old `Linear` has a weight and maybe a bias (which are `Tensor`s).

From what you explained about gradient tracking in Candle, these will not be tracked by default. However, if someone has added them to a `VarBuilder`, they are being tracked. Even if I wrap the `Linear` with a new struct or create a new version of `Linear`, the weights and/or biases could still be tracked, there is nothing I can do.

The reason I need this is because I am implementing LoRA, which is efficient because the old layer is *not* trainable (it is frozen). I need to be able to make sure the gradients from the weights and/or biases of the old `Linear` layer are not being tracked.

It seems to me that the only solution is to add a trainability state to the `Tensor` struct. How can this be implemented, perhaps with a method? Regarding backprop process itself - perhaps any non-trainable `Tensor` (even a variable) would not be added to the DAG? This would ensure they are not treated as variables.

**Comment by LaurentMazare** [comment]
You can call `.detach` on a tensor if you want to be sure that gradient don't go through it, with this you can easily implement `NonTrainableLinear` but maybe I'm missing something?
Also adding the tensor to a `VarBuilder` is not sufficient, you need to add it through a `VarMap` so that it's an actual trainable variable, so as long as you don't do this you should be fine.

**Comment by EricLBuehler** [comment]
Ah, okay! Thank you!
